lxd/cluster/gateway.go

```diff
@@ -196,19 +197,27 @@ func (g *Gateway) HandlerFuncs(nodeRefreshTask func(*APIHeartbeat), trustedCerts
 			}
 
 			// Look for time skews
-			if heartbeatData.Time.Add(5 * time.Second).Before(time.Now().UTC()) {
-				if !g.timeSkew {
-					logger.Warn("Time skew detected between leader and local", log.Ctx{"leaderTime": heartbeatData.Time, "localTime": time.Now().UTC()})
-				}
-				g.timeSkew = true
-			} else if heartbeatData.Time.Add(-5 * time.Second).After(time.Now().UTC()) {
+			now := time.Now().UTC()
+
+			if heartbeatData.Time.Add(5*time.Second).Before(now) || heartbeatData.Time.Add(-5*time.Second).After(now) {
 				if !g.timeSkew {
-					logger.Warn("Time skew detected between leader and local", log.Ctx{"leaderTime": heartbeatData.Time, "localTime": time.Now().UTC()})
+					logger.Warn("Time skew detected between leader and local", log.Ctx{"leaderTime": heartbeatData.Time, "localTime": now})
+
+					err := g.Cluster.UpsertWarning("", "", -1, -1, db.WarningClusterTimeSkew, fmt.Sprintf("leaderTime: %s, localTime: %s", heartbeatData.Time, now))
```

Comment:
@monstermunchkin @stgraber more of a passing though really, but maybe the `message` arg for `UpsertWarning` would be more useful as a structured argument (e.g. a `map[string]string` or a `log.Ctx` type) as it seems to be being used mostly to put contextual info into the warning. We could then encoded that as JSON when its stored in the DB and decode when returned by the API? That would probably mean an API change, but seeing as this feature is relatively new and unused, if that is of any value then now would probably be the easiest time to do it.

Reply:
Sounds like a good idea as the warning type itself (i.e. the string representation) says what it's about.